driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java

```diff
@@ -112,17 +116,28 @@ static BsonDocument getClusterTime(final BsonDocument response) {
         return clusterTime.asDocument();
     }
 
+    static BsonTimestamp getSnapshotTimestamp(final ResponseBuffers responseBuffers) {
+        BsonValue atClusterTimeValue = getNestedFieldValue(responseBuffers, "cursor", "atClusterTime");
+        if (atClusterTimeValue == null) {
+            atClusterTimeValue = getFieldValue(responseBuffers, "atClusterTime");
+        }
+        if (atClusterTimeValue != null && atClusterTimeValue.getBsonType() == BsonType.TIMESTAMP) {
+            return atClusterTimeValue.asTimestamp();
+        }
+        return null;
+    }
+
     static BsonDocument getRecoveryToken(final ResponseBuffers responseBuffers) {
-        return getFieldDocument(responseBuffers, "recoveryToken");
+        BsonValue value = getFieldValue(responseBuffers, "recoveryToken");
+        if (value == null) {
+            return null;
+        }
+        return value.asDocument();
     }
 
-    private static BsonDocument getFieldDocument(final ResponseBuffers responseBuffers, final String fieldName) {
```

Comment:
I'd probably have kept this method just delegated to the getFieldValue method.

Reply:
Agreed.  I added it back, and also did the same for similar timestamp one above